### ISSUE TITLE
Chalice dependency warning when including PyYAML

### ISSUE THREAD

**Issue by jbrow70** [aws/chalice#572]
Hey. We're experiencing some issues including some dependencies in the vendor folder per the instructions on your documentation. The dependency we need is PyYAML 3.12

Our requirements.txt looks like this:
goblin==2.1.0rc2
aiogremlin==3.2.6rc1
gremlinpython==3.2.6
inflection==0.3.1

We needed the PyYAML dependency when we attempted to deploy the application. 
Could not install dependencies:
PyYAML==3.12
You will have to build these yourself and vendor them in
the chalice vendor folder.

So we followed the instructions on the documentation and did a pip download and converted it to a wheel
PyYAML-3.12-cp36-cp36m-linux_x86_64.whl
We extracted the wheel to the vendor folder and got the following structure:
--vendor
     --PyYAML-3.12.dist-info
     --yaml

However, when we run chalice deploy we get a warning. (We ran chalice delete first to check if its clean.) 
Creating role: chalice_clearinghouse-dev
Creating deployment package.

Could not install dependencies:
PyYAML==3.12
You will have to build these yourself and vendor them in
the chalice vendor folder.

Your deployment will continue but may not work correctly
if missing dependencies are not present. For more information:
http://chalice.readthedocs.io/en/latest/topics/packaging.html

/usr/lib/python3.6/zipfile.py:1355: UserWarning: Duplicate name: 'PyYAML-3.12.dist-info/DESCRIPTION.rst'
  return self._open_to_write(zinfo, force_zip64=force_zip64)
/usr/lib/python3.6/zipfile.py:1355: UserWarning: Duplicate name: 'PyYAML-3.12.dist-info/RECORD'
  return self._open_to_write(zinfo, force_zip64=force_zip64)
/usr/lib/python3.6/zipfile.py:1355: UserWarning: Duplicate name: 'PyYAML-3.12.dist-info/METADATA'
  return self._open_to_write(zinfo, force_zip64=force_zip64)
/usr/lib/python3.6/zipfile.py:1355: UserWarning: Duplicate name: 'PyYAML-3.12.dist-info/WHEEL'
  return self._open_to_write(zinfo, force_zip64=force_zip64)
/usr/lib/python3.6/zipfile.py:1355: UserWarning: Duplicate name: 'PyYAML-3.12.dist-info/metadata.json'
  return self._open_to_write(zinfo, force_zip64=force_zip64)
/usr/lib/python3.6/zipfile.py:1355: UserWarning: Duplicate name: 'PyYAML-3.12.dist-info/top_level.txt'
  return self._open_to_write(zinfo, force_zip64=force_zip64)
/usr/lib/python3.6/zipfile.py:1355: UserWarning: Duplicate name: 'yaml/dumper.py'
  return self._open_to_write(zinfo, force_zip64=force_zip64)
/usr/lib/python3.6/zipfile.py:1355: UserWarning: Duplicate name: 'yaml/cyaml.py'
  return self._open_to_write(zinfo, force_zip64=force_zip64)
/usr/lib/python3.6/zipfile.py:1355: UserWarning: Duplicate name: 'yaml/loader.py'
  return self._open_to_write(zinfo, force_zip64=force_zip64)
/usr/lib/python3.6/zipfile.py:1355: UserWarning: Duplicate name: 'yaml/serializer.py'
  return self._open_to_write(zinfo, force_zip64=force_zip64)
/usr/lib/python3.6/zipfile.py:1355: UserWarning: Duplicate name: 'yaml/nodes.py'
  return self._open_to_write(zinfo, force_zip64=force_zip64)
/usr/lib/python3.6/zipfile.py:1355: UserWarning: Duplicate name: 'yaml/error.py'
  return self._open_to_write(zinfo, force_zip64=force_zip64)
/usr/lib/python3.6/zipfile.py:1355: UserWarning: Duplicate name: 'yaml/parser.py'
  return self._open_to_write(zinfo, force_zip64=force_zip64)
/usr/lib/python3.6/zipfile.py:1355: UserWarning: Duplicate name: 'yaml/reader.py'
  return self._open_to_write(zinfo, force_zip64=force_zip64)
/usr/lib/python3.6/zipfile.py:1355: UserWarning: Duplicate name: 'yaml/constructor.py'
  return self._open_to_write(zinfo, force_zip64=force_zip64)
/usr/lib/python3.6/zipfile.py:1355: UserWarning: Duplicate name: 'yaml/emitter.py'
  return self._open_to_write(zinfo, force_zip64=force_zip64)
/usr/lib/python3.6/zipfile.py:1355: UserWarning: Duplicate name: 'yaml/composer.py'
  return self._open_to_write(zinfo, force_zip64=force_zip64)
/usr/lib/python3.6/zipfile.py:1355: UserWarning: Duplicate name: 'yaml/scanner.py'
  return self._open_to_write(zinfo, force_zip64=force_zip64)
/usr/lib/python3.6/zipfile.py:1355: UserWarning: Duplicate name: 'yaml/events.py'
  return self._open_to_write(zinfo, force_zip64=force_zip64)
/usr/lib/python3.6/zipfile.py:1355: UserWarning: Duplicate name: 'yaml/__init__.py'
  return self._open_to_write(zinfo, force_zip64=force_zip64)
/usr/lib/python3.6/zipfile.py:1355: UserWarning: Duplicate name: 'yaml/representer.py'
  return self._open_to_write(zinfo, force_zip64=force_zip64)
/usr/lib/python3.6/zipfile.py:1355: UserWarning: Duplicate name: 'yaml/tokens.py'
  return self._open_to_write(zinfo, force_zip64=force_zip64)
/usr/lib/python3.6/zipfile.py:1355: UserWarning: Duplicate name: 'yaml/resolver.py'
  return self._open_to_write(zinfo, force_zip64=force_zip64)

Our interpreter that we use chalice on also has awscli installed. This package, awscli, already has PyYAML installed as a dependency. My question is, does this warning affect anything upstream when the lambda is deployed? Will lambda have PyYAML? It seems like Chalice mixed up the interpreters dependencies and what we expect.

**Comment by jbrow70** [comment]
Just tested it. It seems that yaml imports correctly. It must be a virtualenv thing. Have you experienced any warnings like this before?

**Comment by JordonPhillips** [comment]
Chalice can't build pyyaml unfortunately. Their `setup.py` fails to detect when using a compiler to compile their C speedups fails, which means that you can't build a proper wheel for it outside of linux. They also don't provide pre-built wheels, which exacerbates the issue.

What you'll need to do is build the wheel on a linux instance (ideally amazon linux), and then vendor that built wheel.

The good news is that there's a fix to the PyYAML `setup.py` that's awaiting release.

**Comment by JordonPhillips** [comment]
If you want to learn more about building the wheels you need, you can see an example [here](http://chalice.readthedocs.io/en/latest/topics/packaging.html#cryptography-example).

As for PyYAML, the commit that fixed the problem I talked about can be found [here](https://github.com/yaml/pyyaml/commit/298e07907ae526594069f6fdf31f2f1278cc1ae3). It's already been merged, so it'll be out in the next release! The current version is `3.12`, so when the fix is release you'll want to make sure you're using `3.13` at least.

Does that answer your question for you?

**Comment by kyleknap** [comment]
Closing due to inactivity.

**Comment by vfilimonov** [comment]
Hello. Just a heads up on Chalice (1.6.0) and PyYAML (3.13).

The `PyYAML==3.13` still needs to be vendored. 
After vendoring procedure and adding `PyYAML==3.13` to `requirements.txt` and deploying the app, all lambda functions are invoked correctly. But during the deployment process the warning is still shown:
```
Creating deployment package.

Could not install dependencies:
pyyaml==3.13
You will have to build these yourself and vendor them in
the chalice vendor folder.

Your deployment will continue but may not work correctly
if missing dependencies are not present. For more information:
http://chalice.readthedocs.io/en/latest/topics/packaging.html

Updating policy for IAM role: .....
```

**Comment by wutachih** [comment]
Was this issue resolved?

I got the same issue lately. I built the wheel on EC2 ubuntu 18.04.

chalice 1.6.2, python 3.7.2, darwin 18.2.0

project folder:
requirements.txt
app.py			chalicelib		vendor

requirements:
PyMySQL
PyJWT>=1.5.3
boto3>=1.9.86
botocore>=1.12.86
kubernetes

vendor folder:
PyYAML-3.13.dist-info	yaml

Creating deployment package.

Could not install dependencies:
pyyaml==3.13
You will have to build these yourself and vendor them in
the chalice vendor folder.

Your deployment will continue but may not work correctly
if missing dependencies are not present. For more information:
http://chalice.readthedocs.io/en/latest/topics/packaging.html
